source/Layer.d

```diff
@@ -63,99 +63,98 @@ abstract class Layer(T)
 
 /+ This layer implement a simple linear matrix transformation
  + applied to the vector.
+ +
+ + M (type): Type of the matrix to be used in the layer.
+ + T (type): Type of the matrix.
+ +
+ + WE ASSUME NOBODY WE'LL USE ANYTHING ELSE THAN A MATRIX FOR "M".
+ + TODO: use some kind of limitation to make sure M is a known matrix
+ + (std.variant, enum, set ...).
+ +
  +/
-class MatrixLayer(T) : Layer!T
+class MatrixLayer(Mtype, T) : Layer!T
 {
-    this(in MatrixAbstract!T _M)
+    this()
+    {
+        // TODO: Add check to see if Mtype is a Matrix*!T
+    }
+
+    this(in Mtype _M)
     {
+        this();
         // We keep a duplicate of the matrix.
         params = new Parameter[1];
         params[0] = _M.dup;
```

Comment:
Can we have a constructor that "Steals" the matrix? perhaps for another pull request, but this sort of copying data as a default is expensive on performance